### DISCUSSION TITLE
js_modules

### DISCUSSION THREAD

**Discussion by WebReflection** [pyscript/pyscript#1873] [Proposals]
Discord is cool, and [an MR about it](https://github.com/pyscript/polyscript/pull/69) is even cooler ... but I want everyone to understand the nitty-gritty behind the scene:

  * anything that could be lazy in PyScript (or polyscript) is lazy to date ... kudos to us going that path!
  * modules landing on the main thread **cannot be lazy** or we would force everyone to use the `async` attribute and `await` in their code
  * current implementation **pre-loads** modules defined in the `js_modules` *py-config* field with the `[files]` *source* -> *target* like approach ... the *target* is the module name to import when/if needed, but it pre-loads module *before* on *main*, just like `packages` field does for pyodide to date ... this is to allow `from thingy import stuff` to always work synchronously in Python code in *main*
  * pre-loaded modules indeed end up as `registerJSModule` operation ... that's the *pyodide* way  to offer to Python code any `from thingy import stuff`
  * when it comes to JS modules, most of them will fail at landing as native JS module in a *worker* ... as most JS modules assume global `window` or `document` out there and won't work with other DOM related classes expected in the module, lazily or not
  * *some* module might work in a *worker* env out of the box and as *worker* purpose is to delegate off main thread anything that's possible to run in a worker, usually also heavy computation, I don't see any reason to force all modules to land on *main* but, most importantly, that cannot be done seamlessly anyway because the interpreter is in a *worker* and we cannot use `registerJSModule` from main there ... or so I suppose (maybe we could, still it would look and feel awkward)
  * with the **hybrid** approach, the *polyscript* logic does the following:
    * it tries to land the module in the current main/worker context, and fail with an error on main, if something bad happens, or it *just warn* in the worker
    * in the *worker*, no matter if that module failed or not as direct import, it will be served through the `xworker.js_modules.thingy` as an automatically proxy thing from the *main* thread

### The Issue

Everything actually works as expected but there's some brain gym to consider:

  * from a *user* perspective, everything should just, and always, work! ... this is unfortunately not possible due the fact we can't know AOT if a module would work in a *worker* or not .... but ideally, any module that could, should be reachable via direct import: `from thingy import stuff` as opposite of `from pyscript.js_module.thingy import stuff` to pass through the worker/main `window` proxy
  * from a technical point of view, we can lazy load anything on demand with workers and also pre-bootstrap modules in the current worker env, but we cannot do the same in a seamless way like we do with `from pyscript import window, document` in the wild ... we need to draw a line, write very explicative docs about it, or simply stop trying to make main/worker worlds looks similar

### Alternative

We go explicit in the `py-config` and instead of `js_module` that tries to apply so much magic with so many possible pitfalls behind scene, we document with ease `js_modules` in a way that stuff can be disambiguated in the code:

```toml
[js_modules.main]
"cdn/main.thing.js" = "main"

[js_modules.worker]
"cdn/worker.thing.js" = "main"
```

This idea would allow users to write code like:

```python
from pyscript.js_modules import main
```

and be sure that wherever that code lands the import will point at the right file for either main or worker and the module would be synchronously available for either main or worker!

#### Technical Caveats / Challenges

As the devil is in the detail, there could be use cases where the module, no matter what, should be landed on the main thread, and the current map example in the mentioned MR is a clear example of that ... my feeling is that modules meant to land on main should be equally reached by workers, *unless explicitly overwritten* so that a config with only *main* modules would work out of the box for both *main* and *worker* case, but one with also *worker* modules should use those for the worker, if present, and fallback to the main module if not ... this should cover 99% of use cases to me, but there is the last chapter to consider ...

### CSS related files

As @ntoll pointed out, it's cool to have also CSS files defined in the *config* and why shouldn't we enable that ... so here are a few options:

  * we disambiguate everything via `[css_modules.main]` and `[css_modules.worker]` options in the config so that all possible use cases should be covered
  * we improve the parser of fields to understand that if `"cdn/main.thing.js" = "main"` is there and `"cdn/main.thing.css" = "main"` is after, but the extension is CSS, it means that these could be defined as modules within the same JSON/TOML field, and magic will happens ...

The last point is:

```toml
[js_modules.main]
"cdn/main.thing.js" = "main"
"cdn/main.thing.css" = "main"

[js_modules.worker]
"cdn/worker.thing.js" = "main"
```

The worker case would infer the `.css` thing is needed by main but I think I am stretching the bar of magic a bit too far ... what do you think instead?

**Comment by WebReflection** [comment]
folks, after sleeping over, I think I'll give last option a chance:

```toml
[js_modules.main]
"cdn/main.thing.js" = "main"
"cdn/main.thing.css" = "main"

[js_modules.worker]
"cdn/worker.thing.js" = "main"
```

I will also try to make this as seamless as possible for users so it will land as `from polyscript import js_modules` and everything should work out of the box

**Comment by Neon22** [comment]
As you point out, it can be difficult for new users to work all of this out. 
How about make a number of initial project wide testing/experimenting setups where:
 - its easy for the user to put their package into a worker (say) and test to see if it works as expected.
 - or in the main, or .. does some more/all the other possibilities
 - I.e. minimal code just to test if this mix will work as expected.
Then the user can see if that package/set of packages will work the way they want - or they will find a layout? that will allow their config to work. They will also be exposed to some different approaches and maybe better get a grasp on how it all fits together.
 
I am guessing there is about four or five main mechanisms ?

**Reply by WebReflection** [reply]
the latest MR makes it deadly simple to normalize expectations at the Python code level and for the config it's a matter of disambiguating main VS worker but main modules work in workers anyway, as explained in the docs:
https://pyscript.github.io/polyscript/#extra-config-features

This is a non-breaking new feature and while there are bits and bops to solve on MicroPython side, the Pyodide story works seamlessly well.

**Comment by WebReflection** [comment]
Closing as this landed already in both polyscript and pyscript as `js_module` config entry.
